### PR TITLE
Add attendance calculator 

### PR DESCRIPTION
**Pull request by mibac138**
| image | image | image | image |
|---|---|---|---|

### CODE REVIEW

**Comment by mklkj** [comment]
@mibac138 zrobisz rebase/merge w wolnej chwili?

File: app/src/main/java/io/github/wulkanowy/data/db/dao/AttendanceSummaryDao.kt [hunk 1]
```diff
@@ -3,6 +3,7 @@ package io.github.wulkanowy.data.db.dao
 import androidx.room.Dao
 import androidx.room.Query
 import io.github.wulkanowy.data.db.entities.AttendanceSummary
+import io.github.wulkanowy.data.db.entities.AttendanceSummaryWithName
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/data/db/dao/AttendanceSummaryDao.kt:6]: remove this import

File: app/src/main/java/io/github/wulkanowy/data/db/entities/AttendanceSummaryWithName.kt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package io.github.wulkanowy.data.db.entities
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/data/db/entities/AttendanceSummaryWithName.kt:1]: remove this file

File: app/src/main/java/io/github/wulkanowy/data/repositories/PreferencesRepository.kt [hunk 3]
```diff
@@ -37,6 +37,12 @@ class PreferencesRepository @Inject constructor(
             R.bool.pref_default_attendance_present
         )
 
+    val targetAttendance: Flow<Int>
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/data/repositories/PreferencesRepository.kt:40]: add "flow" suffix (every other flow-property has it)

File: app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                // Every individual combined flow causes separate network requests to update data.
+                // When there is N child flows, they can cause up to N-1 items to be emitted. Since all
+                // requests are usually completed in less than 5s, there is no need to emit multiple
+                // intermediates that will be visible for barely any time.
+                old is Resource.Intermediate && new is Resource.Intermediate
+            }
+    }
+}
+
+data class AttendanceData(
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt:71]: move this data class to data.pojos

File: app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    val subjectName: String,
+    val allPresences: Int,
+    val allAbsences: Int,
+) {
+    val total: Int
+        get() = allPresences + allAbsences
+
+    val presencePercentage: Double
+        get() = if (total == 0) 0.0
+        else (allPresences.toDouble() / total) * 100
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt:81]: avoid domain-logic in data structure, move those to constructor properties

File: app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        val subjects = subjectRepository.getSubjects(student, semester, forceRefresh)
+            .onResourceError { throw it }.toFirstResult().dataOrNull!!
+
+        return combine(subjects.map { subject ->
+            attendanceSummaryRepository.getAttendanceSummary(
+                student, semester, subject.realId, forceRefresh
+            ).mapResourceData { summaries ->
+                AttendanceData(subject.name,
+                    summaries.sumOf { it.allPresences },
+                    summaries.sumOf { it.allAbsences })
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt:34]: add names to constructor parameters

File: app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        student: Student,
+        semester: Semester,
+        forceRefresh: Boolean,
+    ): Flow<Resource<List<AttendanceData>>> {
+        val subjects = subjectRepository.getSubjects(student, semester, forceRefresh)
+            .onResourceError { throw it }.toFirstResult().dataOrNull!!
+
+        return combine(subjects.map { subject ->
+            attendanceSummaryRepository.getAttendanceSummary(
+                student, semester, subject.realId, forceRefresh
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/domain/attendance/GetAttendanceCalculatorDataUseCase.kt:30]: add names to constructor parameters

File: app/src/main/java/io/github/wulkanowy/ui/modules/attendance/calculator/AttendanceCalculatorAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            val item = items[position]
+            val (name, presences, absences) = item
+            attendanceCalculatorPercentage.text = "${item.presencePercentage.roundToInt()}"
+            // The `+ 1` is to avoid false positives in close cases. Eg.:
+            // target frequency 99%, 1 presence. Without the `+ 1` this would be reported shown as
+            // a positive balance of +1, however that is not actually true as skipping one class
+            // would make it so that the balance would actually be negative (-98). The `+ 1`
+            // fixes this and makes sure that in situations like these, it's not reporting incorrect
+            // balances
+            if (presences / (item.total + 1f) >= targetFreq) {
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/ui/modules/attendance/calculator/AttendanceCalculatorAdapter.kt:41]: soo maybe move this logic to use case?

**Review by mklkj — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/github/wulkanowy/ui/modules/attendance/calculator/AttendanceCalculatorAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import androidx.recyclerview.widget.RecyclerView
+import io.github.wulkanowy.data.pojos.AttendanceData
+import io.github.wulkanowy.databinding.ItemAttendanceCalculatorHeaderBinding
+import javax.inject.Inject
+import kotlin.math.roundToInt
+
+class AttendanceCalculatorAdapter @Inject constructor() :
+    RecyclerView.Adapter<AttendanceCalculatorAdapter.ViewHolder>() {
+
+    var targetFreq: Double = 0.5
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/ui/modules/attendance/calculator/AttendanceCalculatorAdapter.kt:16]: it looks like this is no longer needed

File: app/src/main/java/io/github/wulkanowy/ui/modules/attendance/calculator/AttendanceCalculatorAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                attendanceCalculatorBalanceNeutral.isVisible = true
+                attendanceCalculatorBalanceNegative.isVisible = false
+            }
+            attendanceCalculatorTitle.text = item.subjectName
+            attendanceCalculatorTotal.text = "${item.total}"
+            attendanceCalculatorPresence.text = "${item.presences}"
+        }
+    }
+
+    override fun getItemCount() = items.size
```
Comment (mklkj) [inline app/src/main/java/io/github/wulkanowy/ui/modules/attendance/calculator/AttendanceCalculatorAdapter.kt:56]: move this above onCreateViewHolder() (we always preserve this order in adapters)

**Review by mklkj — CHANGES_REQUESTED** [review]

**Review by Faierbel — APPROVED** [review]

**Review by mklkj — APPROVED** [review]

**Pull request merged**